spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketSession.java

```diff
@@ -51,52 +68,79 @@ public JettyWebSocketSession(Session session, HandshakeInfo info, DataBufferFact
 	public JettyWebSocketSession(Session session, HandshakeInfo info, DataBufferFactory factory,
 			@Nullable Sinks.Empty<Void> completionSink) {
 
-		super(session, ObjectUtils.getIdentityHexString(session), info, factory, completionSink);
-		// TODO: suspend causes failures if invoked at this stage
-		// suspendReceiving();
+		super(session, ObjectUtils.getIdentityHexString(session), info, factory);
+		this.handlerCompletionSink = completionSink;
+		this.flux = Flux.create(emitter -> {
+			this.sink = emitter;
+			emitter.onRequest(n -> {
+				boolean demand = false;
+				this.lock.lock();
+				try {
+					this.requested += n;
+					if (!this.awaitingDemand && this.requested > 0) {
+						this.requested--;
+						this.awaitingDemand = true;
+						demand = true;
+					}
+				}
+				finally {
+					this.lock.unlock();
+				}
+
+				if (demand) {
+					getDelegate().demand();
+				}
+			});
+		});
 	}
 
+	void handleMessage(WebSocketMessage message) {
+		this.sink.next(message);
 
-	@Override
-	protected boolean canSuspendReceiving() {
-		// Jetty 12 TODO: research suspend functionality in Jetty 12
-		return false;
+		boolean demand = false;
+		this.lock.lock();
+		try {
+			if (!this.awaitingDemand) {
+				throw new IllegalStateException();
+			}
+			this.awaitingDemand = false;
+			if (this.requested > 0) {
+				this.requested--;
+				this.awaitingDemand = true;
+				demand = true;
+			}
+		}
+		finally {
+			this.lock.unlock();
+		}
+
+		if (demand) {
+			getDelegate().demand();
+		}
 	}
 
-	@Override
-	protected void suspendReceiving() {
```

Comment:
I very much like that the word "suspend" does not appear in the updates of this PR!

Reply:
It was doing suspend because it was implementing `AbstractListenerWebSocketSession`. 

Implementing directly from `AbstractWebSocketSession` allows us to do the flow control in terms of demand and not a suspend/resume model.